openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/streams/doc/man/man4/loop.4.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) loop.4.man,v 0.9.2.6 2006/04/05 09:43:44 brian Exp
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified 2006/04/05 09:43:44 by brian
.\"
.\" -------------------------------------------------------------------------
.\"
.\" loop.4.man,v
.\" Revision 0.9.2.6  2006/04/05 09:43:44  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database streams.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so streams.macros
.\"
.\"
.TH LOOP 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B loop
\- \fISTREAMS\fP loop-around pseudo-device driver
.\"
.\"
.SH DESCRIPTION
.PP
The
.B loop
psuedo-device multiplexing driver is an implementation of the loop-around driver described in the
.RI \(lq "UNIX\(rg System V Release 4 \- STREAMS Programmer's Guide" ,\(rq
.[
svr4spg
.]
with a number of exceptions and enhancments.  The primary purposes of the
.B loop
pseudo-device driver are: to provide a capability to the
.BR test-streams (8)
conformance and verification test program, for testing passing of STREAMS file descriptors across a
pipe-like driver;
and to provide an example of a pipe-like driver for the STREAM programmer.
To avoid complications for testing, and to serve as a basic example, the driver is as simplistic as
possible.
.PP
The
.B loop
pseudo-device driver has the following behaviour:
.TP 3n
.B "open\fR(2s)"
When a
.B loop
STREAM is opened using
.BR open (2s)
a new
.B loop
STREAM is created.  The newly created STREAM is not associated with any other STREAM.
Issuing the
.B LOOP_SET
input-output control is necessary, to associate the loop STREAM with another one,
before data can be written to or read from the STREAM.  Any attempt to send data on the STREAM
before it is associated with another will result in an
.BR M_ERROR (9)
message being sent upstream causing all further read, write or control operations on the stream to
fail with error
.RB [ ENXIO ].
.TP
.B LOOP_SET
The
.BR LOOP_SET
input-output control described under
.RI \(lq IOCTLS ,\(rq
below, is used to connect the upstream and downstream flow between the requesting STREAM and a
specified STREAM.  Once connected in this fashion, data can be passed in either direction.
.TP
.B "putpmsg\fR(2s), \fPgetpmsg\fR(2s)"
Data is passed in either direction using the
.BR putpmsg (2s)
or
.BR getpmsg (2s)
calls as normal.  The
.B loop
driver coordinates
.I STREAMS
flow control across the driver by explicitly enabling service procedures of the feeding side when
the accepting side is back-enabled.
.TP
.B "close\fR(2s)"
When an STREAM is closed using the
.BR close (2s)
system call, the STREAM is dismantled.
If the upper STREAM remains connected to another STREAM (due to a previous call to
.BR LOOP_SET ),
the connection will be broken and a
.BR M_HANGUP (9)
message will be sent to the other STREAM.
.\"
.\"
.SH USAGE
.PP
The
.B loop
pseudo-device loop-around driver is an implementation of the loop-around driver described in the
.RI \(lq "UNIX\(rg System V Release 4 \- STREAMS Programmer's Guide" ,\(rq
.[
svr4spg
.]
with a number of considerations as follows:
.IP \(bu \w'\(bu\(em'u
The
.I SVR4
driver does not implement input-output controls other than
.BR LOOP_SET .
In contrast,
.B loop
implements a set of input-output controls compatible with the
.I LiS
.[
lis
.]
loop-around driver for use in testing.
.\"
.\"
.SH IOCTLS
.PP
.TP
.B LOOP_SET
.\"
.\"
.SH NOTICES
.PP
The
.B loop
driver is used by the
.BR test-streams (8)
conformance and verification test program to test passing of file descriptors across a
pipe-like driver.  Much of its behaviour is specific to testing and cannot be altered without
impacting the ability of the
.BR test-streams (8)
conformance and verification test program results.
.\"
.\"
.SH IMPLEMENTATION
.PP
.B loop
is implemented as described in the
.IR "SVR4 SPG" .
.[
svr4spg
.]
.\"
.\"
.SH EXAMPLES
.\"
.\"
.SH DEVICES
.PP
.TP
.B /dev/sloop
The external filesystem
.I STREAMS
loop device.
.TP
.BR /dev/streams/loop ", " /devices/loop
The
.BR specfs (5)
shadow special filesystem
.I STREAMS
loop device.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR mux (4),
.BR nullmod (4),
.BR test-streams (8).
.\"
.\"
.SH BUGS
.PP
.B loop
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.B @PACKAGE_TITLE@
.[
lfs
.]
provides the
.B loop
pseudo-device driver for testing and compatibility with
.IR LiS ,
.[
LiS
.]
with the following portability considerations:
.IP \(em \w'\(em\(em'u
Although this driver is described in the
.IR "SVR4 SPG"
.[
svr4spg
.]
as a driver example, most
.I STREAMS
implementations do not actually provide such a driver.
.IP \(em
.B loop
is provided primarily for use by the
.BR test-streams (8)
conformance and verification test program, and as an example of a pipe-like driver with flow
control supported across the driver.  It should not be used by portable
.I STREAMS
modules, driver or applications programs.
.PP
See
.IR STREAMS (9)
for additional compatibility information.
.\"
.\"
.SH CONFORMANCE
.PP
SVR4
.[
svr4spg
.]
and LiS
.[
lis
.]
documentation.
Conformance is verified using the
.BR test-loop (8)
verification test suite.
.\"
.\"
.SH HISTORY
.PP
.B loop
is described as an example loop-around driver in the
.IR "SVR4 SPG" .
.[
svr4spg
.]
.\"
.\"
.[
$LIST$
.]
.TI



openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/streams/src/drivers/loop.c
/*****************************************************************************

 @(#) loop.c,v OpenSS7-0_9_2_D_rc2(0.9.2.8) 2006/06/14 10:37:21

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2005  OpenSS7 Corporation <http://www.openss7.com>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 2 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified 2006/06/14 10:37:21 by brian

 *****************************************************************************/

#ident "@(#) loop.c,v OpenSS7-0_9_2_D_rc2(0.9.2.8) 2006/06/14 10:37:21"

static char const ident[] =
    "loop.c,v OpenSS7-0_9_2_D_rc2(0.9.2.8) 2006/06/14 10:37:21";

/*
 *  This file contains a classic loop driver for SVR 4.2 STREAMS.  The loop driver is a general
 *  purpose STREAMS-based pipe-like driver that can be used for connecting two upper-level STREAMS
 *  together to form a structure similar to a STREAMS-based Pipe, but which has an additional driver
 *  module beneath each STREAM head.  Upper level STREAMS are linked together using ioctl()
 *  commands contained in <sys/loop.h>.
 */
#include <linux/config.h>
#include <linux/version.h>
#include <linux/module.h>
#include <linux/init.h>
#include <linux/sched.h>
#include <linux/interrupt.h>

#include <sys/kmem.h>
#include <sys/stream.h>
#include <sys/strconf.h>
#include <sys/strsubr.h>
#include <sys/ddi.h>
#include <sys/loop.h>

#include "sys/config.h"

#define LOOP_DESCRIP	"UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define LOOP_COPYRIGHT	"Copyright (c) 1997-2005 OpenSS7 Corporation.  All Rights Reserved."
#define LOOP_REVISION	"LfS loop.c,v OpenSS7-0_9_2_D_rc2(0.9.2.8) 2006/06/14 10:37:21"
#define LOOP_DEVICE	"SVR 4.2 STREAMS Null Stream (LOOP) Device"
#define LOOP_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define LOOP_LICENSE	"GPL"
#define LOOP_BANNER	LOOP_DESCRIP	"\n" \
			LOOP_COPYRIGHT	"\n" \
			LOOP_REVISION	"\n" \
			LOOP_DEVICE	"\n" \
			LOOP_CONTACT	"\n"
#define LOOP_SPLASH	LOOP_DEVICE	" - " \
			LOOP_REVISION	"\n"

#ifdef CONFIG_STREAMS_LOOP_MODULE
MODULE_AUTHOR(LOOP_CONTACT);
MODULE_DESCRIPTION(LOOP_DESCRIP);
MODULE_SUPPORTED_DEVICE(LOOP_DEVICE);
MODULE_LICENSE(LOOP_LICENSE);
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-loop");
#endif
#endif

#ifndef CONFIG_STREAMS_LOOP_NAME
#define CONFIG_STREAMS_LOOP_NAME "loop"
#endif
#ifndef CONFIG_STREAMS_LOOP_MODID
#error CONFIG_STREAMS_LOOP_MODID must be defined.
#endif
#ifndef CONFIG_STREAMS_LOOP_MAJOR
#error CONFIG_STREAMS_LOOP_MAJOR must be defined.
#endif

modID_t modid = CONFIG_STREAMS_LOOP_MODID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0);
#endif
MODULE_PARM_DESC(modid, "Module id number for LOOP driver. (0 for auto allocation)");

#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_LOOP_MODID));
MODULE_ALIAS("streams-driver-loop");
#endif

major_t major = CONFIG_STREAMS_LOOP_MAJOR;

#ifndef module_param
MODULE_PARM(major, "h");
#else
module_param(major, uint, 0);
#endif
MODULE_PARM_DESC(major, "Major device number for LOOP driver. (0 for auto allocation)");

#ifdef MODULE_ALIAS
MODULE_ALIAS("char-major-" __stringify(CONFIG_STREAMS_LOOP_MAJOR) "-*");
MODULE_ALIAS("/dev/loop");
#ifdef LFS
MODULE_ALIAS("streams-major-" __stringify(CONFIG_STREAMS_LOOP_MAJOR));
MODULE_ALIAS("/dev/streams/loop");
MODULE_ALIAS("/dev/streams/loop/*");
#endif
#endif

STATIC struct module_info loop_minfo = {
	.mi_idnum = CONFIG_STREAMS_LOOP_MODID,
	.mi_idname = CONFIG_STREAMS_LOOP_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

typedef struct loop {
	struct loop *next;		/* list linkage */
	struct loop **prev;		/* list linkage */
	struct loop *other;		/* other stream */
	dev_t dev;			/* device number */
	queue_t *rq;			/* this rd queue */
	queue_t *wq;			/* this wr queue */
} loop_t;

STATIC struct loop *loop_opens = NULL;
STATIC spinlock_t loop_lock = SPIN_LOCK_UNLOCKED;

STATIC streamscall int
loop_wput(queue_t *q, mblk_t *mp)
{
	struct loop *p = q->q_ptr, *o;
	int err;

	switch (mp->b_datap->db_type) {
	case M_IOCTL:
	{
		union ioctypes *ioc = (typeof(ioc)) mp->b_rptr;

		switch (ioc->iocblk.ioc_cmd) {
		case LOOP_SET:
		{
			dev_t dev;

			if (ioc->iocblk.ioc_count != sizeof(dev)) {
				err = -EINVAL;
				break;
			}
			if (mp->b_cont == NULL) {
				err = -EINVAL;
				break;
			}
			dev = *(dev_t *) mp->b_cont->b_rptr;
			spin_lock_bh(&loop_lock);
			for (o = loop_opens; o; o = o->next)
				if (o->dev == dev)
					break;
			if (o == NULL) {
				spin_unlock_bh(&loop_lock);
				err = -ENXIO;
				break;
			}
			if (p->other || o->other) {
				spin_unlock_bh(&loop_lock);
				err = -EBUSY;
				break;
			}
			p->other = o;
			o->other = p;
			enableok(o->wq);
			enableok(p->wq);
			qenable(o->wq);
			qenable(p->wq);
			spin_unlock_bh(&loop_lock);
			err = 0;
			break;
		}
		default:
			err = -EINVAL;
			break;
		}
		if (err == 0) {
			mp->b_datap->db_type = M_IOCACK;
			ioc->iocblk.ioc_count = 0;
			ioc->iocblk.ioc_rval = 0;
			ioc->iocblk.ioc_error = 0;
		} else {
			mp->b_datap->db_type = M_IOCNAK;
			ioc->iocblk.ioc_count = 0;
			ioc->iocblk.ioc_rval = -1;
			ioc->iocblk.ioc_error = -err;
		}
		qreply(q, mp);
		break;
	}
	case M_FLUSH:
		if (mp->b_rptr[0] & FLUSHW) {
			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(q, mp->b_rptr[1], FLUSHALL);
			else
				flushq(q, FLUSHALL);
		}
		spin_lock_bh(&loop_lock);
		if (p->other) {
			queue_t *rq;

			rq = p->other->rq;
			spin_unlock_bh(&loop_lock);
			/* switch sense of flush flags */
			switch (mp->b_rptr[0] & (FLUSHW | FLUSHR | FLUSHRW)) {
			case FLUSHR:
				mp->b_rptr[0] = (mp->b_rptr[0] & ~FLUSHR) | FLUSHW;
				break;
			case FLUSHW:
				mp->b_rptr[0] = (mp->b_rptr[0] & ~FLUSHW) | FLUSHR;
				break;
			}
			putnext(rq, mp);
		} else {
			spin_unlock_bh(&loop_lock);
			if (mp->b_rptr[0] & FLUSHR) {
				if (mp->b_rptr[0] & FLUSHBAND)
					flushband(RD(q), mp->b_rptr[1], FLUSHALL);
				else
					flushq(RD(q), FLUSHALL);
				mp->b_rptr[0] &= ~FLUSHW;
				qreply(q, mp);
			} else {
				freemsg(mp);
			}
		}
		break;
	default:
		spin_lock_bh(&loop_lock);
		if (p->other) {
			if (!q->q_first) {
				queue_t *rq;

				rq = p->other->rq;
				spin_unlock_bh(&loop_lock);
				if (rq) {
					if (mp->b_datap->db_type >= QPCTL
					    || bcanputnext(rq, mp->b_band)) {
						putnext(rq, mp);
						break;
					}
				}
			} else
				spin_unlock_bh(&loop_lock);
			putq(q, mp);
		} else {
			spin_unlock_bh(&loop_lock);
			freemsg(mp);
			putnextctl2(OTHERQ(q), M_ERROR, ENXIO, ENXIO);
		}
		break;
	}
	return (0);
}

/*
 *  The driver only puts messages on its write queue when they fail flow control on the queue beyond
 *  the other read queue.  When flow control subsides, the read service procedure on the othe STREAM
 *  will explicitly enable us.
 */
STATIC streamscall int
loop_wsrv(queue_t *q)
{
	struct loop *p = q->q_ptr;
	queue_t *rq = NULL;
	mblk_t *mp;

	spin_lock_bh(&loop_lock);
	if (p->other)
		rq = p->other->rq;
	spin_unlock_bh(&loop_lock);

	if (rq) {
		while ((mp = getq(q))) {
			if (mp->b_datap->db_type >= QPCTL || bcanputnext(rq, mp->b_band)) {
				putnext(rq, mp);
				continue;
			}
			putbq(q, mp);
			break;
		}
	} else
		flushq(q, FLUSHALL);

	return (0);
}

/*
 *  The driver never places messages on the read queue, therefore, the read service procedure is
 *  only invoked when backenabled from upstream.  Back enabling is due to flow control that caused a
 *  message to tbe put (back) to the other STREAM's write queue so the driver needs to explicitly
 *  enable that write queue now.
 */
STATIC streamscall int
loop_rsrv(queue_t *q)
{
	struct loop *p = q->q_ptr;

	spin_lock_bh(&loop_lock);
	if (p->other && p->other->wq)
		qenable(p->other->wq);
	spin_unlock_bh(&loop_lock);
	return (0);
}

/* 
 *  -------------------------------------------------------------------------
 *
 *  OPEN and CLOSE
 *
 *  -------------------------------------------------------------------------
 */
STATIC streamscall int
loop_open(queue_t *q, dev_t *devp, int oflag, int sflag, cred_t *crp)
{
	struct loop *p, **pp = &loop_opens;
	major_t cmajor = getmajor(*devp);
	minor_t cminor = getminor(*devp);

	_ptrace(("%s: opening major %hu, minor %hu, sflag %d\n", __FUNCTION__, cmajor, cminor,
		sflag));
	if (q->q_ptr != NULL) {
		_printd(("%s: stream is already open\n", __FUNCTION__));
		return (0);	/* already open */
	}
	if (sflag == MODOPEN || WR(q)->q_next) {
		_printd(("%s: cannot open as module\n", __FUNCTION__));
		return (ENXIO);	/* can't open as module */
	}
	if (!(p = kmem_alloc(sizeof(*p), KM_NOSLEEP))) {	/* we could sleep */
		_printd(("%s: could not allocate private structure\n", __FUNCTION__));
		return (ENOMEM);	/* no memory */
	}
	bzero(p, sizeof(*p));
	switch (sflag) {
	case CLONEOPEN:
		_printd(("%s: clone open\n", __FUNCTION__));
		if (cminor < 1)
			cminor = 1;
	case DRVOPEN:
	{
		major_t dmajor = cmajor;

		_printd(("%s: driver open\n", __FUNCTION__));
		if (cminor < 1) {
			_printd(("%s: attempt to open minor zero non-clone\n", __FUNCTION__));
			return (ENXIO);
		}
		spin_lock_bh(&loop_lock);
		for (; *pp && (dmajor = getmajor((*pp)->dev)) < cmajor; pp = &(*pp)->next) ;
		for (; *pp && dmajor == getmajor((*pp)->dev) &&
		     getminor(makedevice(cmajor, cminor)) != 0; pp = &(*pp)->next) {
			minor_t dminor = getminor((*pp)->dev);

			if (cminor < dminor)
				break;
			if (cminor == dminor) {
				if (sflag == CLONEOPEN)
					cminor++;
				else {
					spin_unlock_bh(&loop_lock);
					kmem_free(p, sizeof(*p));
					pswerr(("%s: stream already open!\n", __FUNCTION__));
					return (EIO);	/* bad error */
				}
			}
		}
		if (getminor(makedevice(cmajor, cminor)) == 0) {	/* no minors left */
			spin_unlock_bh(&loop_lock);
			kmem_free(p, sizeof(*p));
			_printd(("%s: no minor devices left\n", __FUNCTION__));
			return (EBUSY);	/* no minors left */
		}
		p->dev = *devp = makedevice(cmajor, cminor);
		p->rq = q;
		p->wq = WR(q);
		noenable(p->wq);
		if ((p->next = *pp))
			p->next->prev = &p->next;
		p->prev = pp;
		*pp = p;
		q->q_ptr = WR(q)->q_ptr = p;
		spin_unlock_bh(&loop_lock);
		qprocson(q);
		_printd(("%s: opened major %hu, minor %hu\n", __FUNCTION__, cmajor, cminor));
		return (0);
	}
	}
	pswerr(("%s: bad sflag %d\n", __FUNCTION__, sflag));
	return (ENXIO);
}

STATIC streamscall int
loop_close(queue_t *q, int oflag, cred_t *crp)
{
	struct loop *p;

	_trace();
	if ((p = q->q_ptr) == NULL) {
		pswerr(("%s: already closed\n", __FUNCTION__));
		return (0);	/* already closed */
	}
	qprocsoff(q);
	spin_lock_bh(&loop_lock);
	if (p->other && p->other->rq)
		putnextctl(p->other->rq, M_HANGUP);
	if ((*(p->prev) = p->next))
		p->next->prev = p->prev;
	p->next = NULL;
	p->prev = &p->next;
	p->other = NULL;
	p->dev = 0;
	p->rq = NULL;
	p->wq = NULL;
	q->q_ptr = WR(q)->q_ptr = NULL;
	spin_unlock_bh(&loop_lock);
	_printd(("%s: closed stream with read queue %p\n", __FUNCTION__, q));
	return (0);
}

STATIC struct qinit loop_rqinit = {
	.qi_srvp = loop_rsrv,
	.qi_qopen = loop_open,
	.qi_qclose = loop_close,
	.qi_minfo = &loop_minfo,
};

STATIC struct qinit loop_wqinit = {
	.qi_putp = loop_wput,
	.qi_srvp = loop_wsrv,
	.qi_minfo = &loop_minfo,
};

STATIC struct streamtab loop_info = {
	.st_rdinit = &loop_rqinit,
	.st_wrinit = &loop_wqinit,
};

STATIC struct cdevsw loop_cdev = {
	.d_name = CONFIG_STREAMS_LOOP_NAME,
	.d_str = &loop_info,
	.d_flag = D_MP,
	.d_fop = NULL,
	.d_mode = S_IFCHR | S_IRUGO | S_IWUGO,
	.d_kmod = THIS_MODULE,
};

#ifdef CONFIG_STREAMS_LOOP_MODULE
STATIC
#endif
int __init
loop_init(void)
{
	int err;

#ifdef CONFIG_STREAMS_LOOP_MODULE
	printk(KERN_INFO LOOP_BANNER);
#else
	printk(KERN_INFO LOOP_SPLASH);
#endif
	loop_minfo.mi_idnum = modid;
	if ((err = register_strdev(&loop_cdev, major)) < 0)
		return (err);
	if (major == 0 && err > 0)
		major = err;
	return (0);
};

#ifdef CONFIG_STREAMS_LOOP_MODULE
STATIC
#endif
void __exit
loop_exit(void)
{
	unregister_strdev(&loop_cdev, major);
};

#ifdef CONFIG_STREAMS_LOOP_MODULE
module_init(loop_init);
module_exit(loop_exit);
#endif
